octobot_trading/traders/trader.py

```diff
@@ -145,48 +154,122 @@ async def _create_not_loaded_order(self, new_order: Order, portfolio) -> Order:
+                if raw_cancelled_order is not None:
+                    await self.exchange_manager.exchange_personal_data.handle_order_update_from_raw(order.symbol,
+                                                                                                    order.order_id,
+                                                                                                    raw_cancelled_order)
+
+    async def _handle_order_cancellation(self, order: Order, is_cancelled_from_exchange: bool) -> dict:
+        raw_cancelled_order = None
+        # if real order: cancel on exchange
+        if not self.simulate and not order.is_self_managed() and not is_cancelled_from_exchange:
+            raw_cancelled_order = await self.exchange_manager.exchange.cancel_order(order.order_id, order.symbol)
+
+        # add to trade history and notify
+        await self.exchange_manager.exchange_personal_data.handle_trade_instance_update(
+            create_trade_from_order(order,
+                                    close_status=OrderStatus.CANCELED,
+                                    canceled_time=self.exchange_manager.exchange.get_exchange_current_time()))
+
+        # update portfolio with cancelled funds order
+        async with self.exchange_manager.exchange_personal_data.get_order_portfolio(order).lock:
+            self.exchange_manager.exchange_personal_data.get_order_portfolio(order) \
+                .update_portfolio_available(order, is_new_order=False)
+
+        if order.is_self_managed():
+            # remove order from open_orders
+            self.exchange_manager.exchange_personal_data.orders_manager.remove_order_instance(order)
+
+        return raw_cancelled_order
 
     async def cancel_order_with_id(self, order_id):
+        """
+        Gets order matching order_id from the OrderManager and calls self.cancel_order() on it
+        :param order_id: Id of the order to cancel
+        :return: True if cancel is successful, False if order is not found
+        """
         try:
             await self.cancel_order(self.exchange_manager.exchange_personal_data.orders_manager.get_order(order_id))
             return True
         except KeyError:
             return False
 
-    # Should be called only if we want to cancel all symbol open orders (no filled)
     async def cancel_open_orders(self, symbol, cancel_loaded_orders=True):
-        # use a copy of the list (not the reference)
-        for order in copy.copy(self.exchange_manager.exchange_personal_data.orders_manager.get_open_orders()):
+        """
+        Should be called only if the goal is to cancel all open orders for a given symbol
+        :param symbol: The symbol to cancel all orders on
+        :param cancel_loaded_orders: When True, also cancels loaded orders (order that are not from this bot instance)
+        :return: None
+        """
+        for order in self.exchange_manager.exchange_personal_data.orders_manager.get_open_orders():
```

Comment:
Why not iterating on copy of `get_open_orders()` anymore ?

Reply:
get_open_orders() already sends a new list, it's not the underlying order dict anymore in v4

Reply:
Oh yes you're right :+1: